### ISSUE TITLE
Feature request: Implementing spatially-sparse conv networks in TensorFlow

### ISSUE THREAD

**Issue by jstaker7** [tensorflow/tensorflow#1604]
I am inspired by Dr. Ben Graham's recent work regarding spatially-sparse convolutional neural networks:

http://arxiv.org/abs/1409.6070 (particularly section 2.3)
http://www2.warwick.ac.uk/fac/sci/statistics/staff/academic-research/graham/sparse3d.pdf

He has graciously open-sourced his neural network library but I'd like to utilize these same ideas in my models, which are implemented in Tensorflow.

Unfortunately, I'm finding his implementation and citations a little hard to follow, but his description of a "feature matrix" and a "pointer matrix" sounds a little like something that can be implemented using sparse variable updates in Tensorflow (https://www.tensorflow.org/versions/r0.7/api_docs/python/state_ops.html#sparse-variable-updates), but I'm afraid that it might not be so simple and would require building custom kernels to support a new-ish type of convolution. But I don't know enough to say for certain which direction to take.

Any thoughts on how we can bring spatially-sparse convolutions to Tensorflow? Anyone interested in collaborating on implementing this?

**Comment by girving** [comment]
Yes, this sounds like it requires custom kernels to have reasonable performance, though I don't see the word "pointer" in the paper so I can't confirm.

In terms of collaboration, do you want to post on the discuss@tensorflow list asking if anyone wants to collaborate?  It's a good use of the list, and most people outside the team are unlikely to notice a Github issue.

Note that we are doing some work on better sparse ops (e.g., @ebrevdo's https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/kernels/sparse_tensor_dense_matmul_op.cc), but I don't think sparse convolutions is on our radar just yet.  It would be a good thing to have, and I'd be happy to answer questions if you do end up working on it.

**Comment by jstaker7** [comment]
Thanks, Geoffrey, that is a great idea, I'll go ahead and post to the email thread. I am definitely interested in implementing this since I am currently working with a sparse dataset (and maybe more in the future), although I don't currently have the skill set to implement it on my own. Thanks for the offer, I'll definitely come back with questions as they come up.

**Comment by iurilgit** [comment]
hi, 
We are interested in the sparse tensor for conv2d as well. Is it supported in the latest version yet?
Thanks!

**Comment by evanfeinberg** [comment]
Just wondering, did anything ever come of this? It would be a real boon to the community!

**Comment by girving** [comment]
Not that I know of.  Still contributions welcome.

**Comment by evanfeinberg** [comment]
Thanks! My thesis is a bit far ATM from this sort of core algorithm development but it would be fun to get into this as a side project if I have the chance.

**Comment by damro** [comment]
What's the status of this? Anyone working on it already?

**Comment by thangduong** [comment]
+1.  I am also interested in this!

**Comment by etienne87** [comment]
There is now implementation in torch/ pytorch: https://github.com/facebookresearch/SparseConvNet.

**Comment by mushroom1116** [comment]
I'm trying to implementing the main idea of SparseCNN described in https://arxiv.org/pdf/1505.02890.pdf (page5) and encountered some problems. Q and H of each layer are different. In order to get the value of Q and H_out, I have to use sess.run() and it may cut the connection of gradient variables. 

I wonder if there are some bottlenecks in implementing SparseCNN using tensorflow API.

**Comment by ebrevdo** [comment]
We'd have to see more of your code, or better yet a design proposal for how
you want to implement the algorithm in tf.  Your options for avoiding
multiple session run calls include generating q and h in tf directly, out
using py_func.

**Comment by mushroom1116** [comment]
@ebrevdo  I want to design a CNN Net with two layers and there are some problems in back propagation. When I got M_out1(M_out1 = (Q1*W)+B) as the input for the second layer, I have to iterate H_out1 of the first layer to determine the H_out2 and Q2 of the second layer.  

Now I access the active site's coordinate one bye one(for i in range(a_out1), a_out1 must be an int and need sess.run()) stored in H_out1 combined with filter size to infer the active site's coordinate in H_out2. 

It seems not a linear reversible transformation and compute Q from M_in need coordinate of active sites and a_in(number of active sites) who need to be run out by session, and it can't back propagate. How to solve this problem? Thank you very much for your time.

**Comment by ghexp** [comment]
Can the custom ops from this approach be brought in under contrib to help address this issue?
https://github.com/uber/sbnet

**Comment by IdRatherBeCoding** [comment]
I implemented the sparse CNN (forward prop only) in python as a learning exercise, computing Q and H with py_func. Comments welcome: https://github.com/IdRatherBeCoding/sparse_cnn

**Comment by mushroom1116** [comment]
@IdRatherBeCoding You just implemented the forward propagate, right?

**Comment by IdRatherBeCoding** [comment]
@mushroom1116 Actually, I recently started work on the gradient, and I'm almost done.

Do you need an efficient sparse CNN in TF for a project, or are you just curious?

To be clear, both my forward and back prop implementations are going to be slower than TF conv2d because I'm computing Q with simple python loops. I intend to write the Op in C++, but it's still just for a learning exercise. Support for maxpooling and strides > 1 will also be needed before it's useful.

**Comment by mushroom1116** [comment]
@IdRatherBeCoding I tried to implement SparseCNN too and I was stopped by some difficulties in BP, so I want to know how  you did it. If may. Could you please let me know the details of code?

**Comment by IdRatherBeCoding** [comment]
@mushroom1116 You can take a look at it now:
https://github.com/IdRatherBeCoding/sparse_cnn/blob/master/sparse_cnn_back_prop.ipynb

Please let me know if something doesn't make sense.

I have tested the derivatives of Q with respect to M and ground state. I have not yet tested with an end to end CNN training.

**Comment by harshini-gadige** [comment]
@IdRatherBeCoding Hi, any update on your work ? I can close this issue if your contribution would suffice the request.

**Comment by asdfzxh8** [comment]
No such thing as enough or fun or not about it. cpu, Can say any no matter what.

**Comment by harshini-gadige** [comment]
Closing this issue for now. Feel free to reopen and share the PR number if any contributions are made.

**Comment by IdRatherBeCoding** [comment]
@harshini-gadige no I didn't pursue this further. As mentioned by @ghexp the sbnets (https://github.com/uber/sbnet) provides similar solution, with better speedups, according to the paper.
